rfd/0172-tbot-ssh-muxer.md

```diff
@@ -0,0 +1,171 @@
+   port templated into it by OpenSSH with the `%h` and `%p` placeholders.
+
+Comparing Rust and Go for this lightweight binary:
+
+- Rust has a smaller binary size: 400KiB vs 1.8MiB.
+- Rust executes faster.
+- Tuning Go's GC behaviour requires also using `env` as part of the
+  ProxyCommand. This is not necessary with Rust.
+
+For this binary, we will use Rust.
+
+### Initial Scope
+
+For the initial version, we must include:
+
+- The long-lived proxying background service within `tbot`
+- The new lightweight client binary with support for ProxyUseFdPass.
+- A configuration option to control the maximum number of proxied connections.
+  - When this maximum is reached, the service will wait until the number of
+    active connections drops below the maximum.
+  - This will allow a sane limit to be provided to prevent the long-lived `tbot`
+    process from being overwhelmed.
+- Prometheus metrics which exposes the health of the proxying.
+  - Number of active proxied connections.
+  - Number of attempts to establish proxied connections, the time this has
+    taken, and the success status.
+- OpenTelemetry trace spans at key points.
+  - This will allow us to target future performance improvements.
+
+### Future Scope
+
+#### Caching
+
+We can further improve performance by caching the results of operations such as
+resolving the target host from labels.
+
+This work is deferred as it is complex and introduces risks. Information from
+OpenTelemetry tracing will allow us to target the most impactful areas at a 
+later date.
+
+#### Upstream Connection Pooling
+
+The initial version will not share the upstream connections to the Teleport
+Proxy.
+
+Sharing upstream connections will allow us to reduce the number of connections
+and reduce the latency and overhead associated with establishing a new
+connection.
+
+This work is deferred as it is complex and introduces risks. We should first
+ensure that the initial version is stable and to take measurements which can
+inform the design of the connection pooling.
+
+### UX
+
+The user will configure `ssh-local-proxy` in the `tbot` configuration file:
+
+```yaml
+services:
+  - type: ssh-local-proxy
+    # Destination is where the ssh_config and UDS will be written to. It must be
+    # a directory.
+    destination:
+      type: directory
+      path: /opt/machine-id
+    # The maximum number of connections that can be proxied at once.
+    max_connections: 100
+```
+
+### Security
+
+#### Unix Domain Socket is unauthenticated
+
+The Unix Domain Socket will be unauthenticated, and this means that any process
+which is able to connect to it will be able to establish upstream connections.
+
+Fortunately, existing Linux permissions mechanisms can be used to restrict 
+access to the UDS and this is similar to how we handle the credentials
+output by `tbot` today.
+
+We should ensure we document the best practices to secure the UDS.
```

Comment:
Permissions on unix sockets are always kind of weird and there's no good way to set the permission of a socket as you're creating it except by playing with the umask (which is process-wide, yuck), so we should take care to use a _subdirectory_ with restrictive permissions that contains the socket - at which point the socket permissions don't matter beyond being at least `u+rw`.

Reply:
At least for `tbot`, we typically operate on the assumption that the user the bot runs as will often not be the same user that is consuming the credentials output by `tbot`. We handled this for the Workload Identity API by chmodding the newly created socket ( which obviously is a little racy) to 777 and then encouraging control by placing more restrictive permissions on the directory which is not recreated on each start.

Reply:
The listening socket should still be in a dedicated subdirectory that we chmod 700 before doing anything else.

Reply:
I'm fine with using `tbot`'s directory ACL management